In [139]:
# Initialize OK
from client.api.notebook import Notebook
ok = Notebook('proj1a.ok')

Assignment: proj1a
OK, version v1.13.11



# Project 1A: Food Safety
## Cleaning and Exploring Data with Pandas
## Due Date: Monday 02/17, 11:59 PM
## Collaboration Policy

Data science is a collaborative activity. While you may talk with others about
the project, we ask that you **write your solutions individually**. If you do
discuss the assignments with others please **include their names** at the top
of your notebook.

**Collaborators**: *I worked alone.*


## This Assignment

In this project, we will investigate restaurant food safety scores for restaurants in San Francisco. The scores and violation information have been [made available by the San Francisco Department of Public Health](https://data.sfgov.org/Health-and-Social-Services/Restaurant-Scores-LIVES-Standard/pyih-qa8i). The main goal for this assignment is to walk through the process of Data Cleaning and EDA. 

We have broken this two week project into two parts.  The first part (week 1) will focus on basic data cleaning and analysis using Pandas.  The second part (week 2) will dive into deeper analysis and data visualization.

As we clean and explore these data, you will gain practice with:
* Reading simple csv files and using Pandas
* Working with data at different levels of granularity
* Identifying the type of data collected, missing values, anomalies, etc.
* Exploring characteristics and distributions of individual variables

## Score Breakdown 
Question | Points
--- | ---
1a | 3
1b | 2
1c | 1
1d | 1
1e | 1
2a | 1
2b | 1
2ci | 1
2cii | 1
2d | 2
3a | 2
3b | 2
3ci | 1
3cii | 1
3d | 3
3e | 2
3f | 2
4a | 1
4bi | 1
4bii | 1
4biii | 1
4ci | 1
4cii | 1
4ciii | 1
4civ | 1
4di | 2
4dii | 2
4e | 2
5a | 2
5b | 3
5c | 2
6 | 3
Total | 51

## Before You Start

For each question in the assignment, please write down your answer in the answer cell(s) right below the question. 

We understand that it is helpful to have extra cells breaking down the process towards reaching your final answer. If you happen to create new cells below your answer to run codes, **NEVER** add cells between a question cell and the answer cell below it. It will cause errors when we run the autograder, and it will sometimes cause a failure to generate the PDF file.

**Important note: The local autograder tests will not be comprehensive. You can pass the automated tests in your notebook but still fail tests in the autograder.** Please be sure to check your results carefully.

Finally, unless we state otherwise, try to avoid using python for loops or list comprehensions.  The majority of this assignment can be done using builtin commands in Pandas and numpy.  


In [140]:
import numpy as np
import pandas as pd 

import bz2 # Used to read compressed data
import os # Used to interact with the file system

## Obtaining the Data

### File Systems and I/O



In general, we will focus on using python commands to investigate files.  However, it can sometimes be easier to use shell commands in your local operating system.  The following cells demonstrate how to do this.


The command below starts with `!`. This tells our Jupyter notebook to pass this command to the operating system. In this case, the command is the `ls` POSIX command which lists all files in the current directory. Note what `!ls data` outputs.  

*Note this `!` only works in ipython shells (Jupyter Notebooks). And the `ls` (list) command only works in posix environments and may not work on default Windows systems.*

In [141]:
!ls 

data  proj1a.ipynb  proj1a.ok  tests


In [142]:
!ls data

bus.csv.bz2  ins2vio.csv.bz2  ins.csv.bz2  sf_zipcodes.json  vio.csv.bz2


We are going to use the [`pathlib`](https://docs.python.org/3/library/pathlib.html#basic-use) module to represent our file system paths and perform operations which allow us to learn more about the contents of our data. Note what `pathlib.Path.cwd()` outputs in relation to the output of `!ls` above.

<br/><br/><br/>

---

<br/><br/><br/>

# 1: Examining the Files
Let's first focus on understanding the structure of the data; this involves answering questions such as:

* How much data do we have?
* Is the data in a standard format or encoding?
* Is the data organized in records?
* What are the fields in each record?

Let's start by looking at the contents of `data/`. This is not just a single file but rather a directory of multiple compressed files. We could inspect our data by uncompressing each file but in this project we're going to do almost everything in Python for maximum portability.


You will want to use a few useful python functions.  To move through the local filesystem you can use the `Path` module in `pathlinb`.  For example, to list the current directory you can [Path.cwd](https://docs.python.org/3/library/pathlib.html#pathlib.Path.cwd).

In [143]:
from pathlib import Path

Path.cwd()

PosixPath('/home/jovyan/sp20/proj/proj1a')

The function returns a `pathlib.Path` object representing the location of the file.  It can also be used to list contents of directories and many other things. 

You will also need to work with `bzip2` files and you will want to be able to read their contents using the [bz2](https://docs.python.org/3/library/bz2.html) python library.

In [144]:
with bz2.open("data/bus.csv.bz2", "r") as f:
    print("The first line:", "\n\t", f.readline())

The first line: 
	 b'"business id column","name","address","city","state","postal_code","latitude","longitude","phone_number"\n'


<br/><br/><br/>

---

## Question 1a:

Implement the `list_files`, `get_file_size`, and `get_linecount_bz2` functions to return the list of files in the directory, the sizes (in bytes) of a file, and the number of lines in the file.  Note the last `get_linecount_bz2` should not produce any intermediate files in the filesystem and should avoid storing the entire file in memory (don't do `len(file.readlines())`).

**Hints:** *You might find the following documentation useful:*
1. [Python pathlib](https://docs.python.org/3.7/library/pathlib.html#pathlib.Path.iterdir)
1. [bz2](https://docs.python.org/3/library/bz2.html#bz2.open)


<!--
BEGIN QUESTION
name: q1a
points: 3
-->

In [145]:
def list_files(directory):
    """
    Return a list of pathlib.Path objects for the files in the directory.
    
    directory: a string describing the directory to list 
        for example 'data/'
    """
    p = Path(directory)
    return [f for f in p.iterdir()]

def get_file_size(file_name):
    """
    Return file size for a given filename.
    """ 
    return os.stat(file_name).st_size

def get_linecount_bz2(file_name):
    """
    Returns the number of lines in bz2 file.  
    """ 
    return sum([1 for line in bz2.open(file_name)])

In [146]:
ok.grade("q1a");

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Running tests

---------------------------------------------------------------------
Test summary
    Passed: 3
    Failed: 0
[ooooooooook] 100.0% passed



Now, let's see the file size and the number of lines for each data file.  If you implemented the above code correctly the following cell should produce the following (the columns may be in a different order):

<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
       <th>linecount</th>
      <th>name</th>
      <th>size</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>2</th>
       <td>66</td>
      <td>data/vio.csv.bz2</td>
      <td>1337</td>
    </tr>
    <tr>
      <th>3</th>
      <td>26664</td>
      <td>data/ins.csv.bz2</td>
      <td>110843</td>
    </tr>
    <tr>
      <th>0</th>
      <td>6254</td>
      <td>data/bus.csv.bz2</td>
      <td>113522</td>
    </tr>
    <tr>
      <th>1</th>
      <td>40211</td>
      <td>data/ins2vio.csv.bz2</td>
      <td>146937</td>
    </tr>
  </tbody>
</table>

In [147]:
info = []
for f in list_files("data/"):
    name = str(f)
    if name[-3:] == "bz2": 
        size = get_file_size(f)
        linecount = get_linecount_bz2(f)
        info.append({"name": name, "size": size, "linecount": linecount})

file_info = pd.DataFrame(info).sort_values("size")
file_info

name    size  linecount
3      data/vio.csv.bz2    1337         66
1      data/ins.csv.bz2  110843      26664
2      data/bus.csv.bz2  113522       6254
0  data/ins2vio.csv.bz2  146937      40211

<br/><br/><br/>

---

## Question 1b: Programatically Looking Inside the Files

Implement the following function `head_bz2` to return a list of the first `nlines` lines of each file. 
Using your `head_bz2` function implement the following `print_head_bz2` function that uses `print()` to print the filename followed by the first `nlines` of each file and their line numbers in the following format.

```
data/bus.csv.bz2
0 :	 b'"business id column","name","address","city","state","postal_code","latitude","longitude","phone_number"\n'
1 :	 b'"1000","HEUNG YUEN RESTAURANT","3279 22nd St","San Francisco","CA","94110","37.755282","-122.420493","-9999"\n'
2 :	 b'"100010","ILLY CAFFE SF_PIER 39","PIER 39  K-106-B","San Francisco","CA","94133","-9999","-9999","+14154827284"\n'
3 :	 b'"100017","AMICI\'S EAST COAST PIZZERIA","475 06th St","San Francisco","CA","94103","-9999","-9999","+14155279839"\n'
4 :	 b'"100026","LOCAL CATERING","1566 CARROLL AVE","San Francisco","CA","94124","-9999","-9999","+14155860315"\n'
```

Do not read the entire file contents!

<!--
BEGIN QUESTION
name: q1b
points: 2
-->

In [148]:
def head_bz2(file, nlines=5):
    """
    Return a list of the first nlines lines of filename
    """
    with bz2.open(file) as f:
        five_head = [f.readline() for _ in range(nlines)]
    return five_head

def print_head_bz2(file, nlines=5):
    """
    Print a list of the first nlines lines of filename
    """
    print(str(file))
    head_list = head_bz2(file, nlines)
    for i in range(nlines):
        print(i, ': ', head_list[i])

In [149]:
ok.grade("q1b");

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Running tests

---------------------------------------------------------------------
Test summary
    Passed: 2
    Failed: 0
[ooooooooook] 100.0% passed



The following should display the filename and head of the file for all the files in data:

In [150]:
for file in list_files("data/"):
    if str(file)[-3:] == "bz2":  
        print_head_bz2(file)
        print()

data/ins2vio.csv.bz2
0 :  b'"iid","vid"\n'
1 :  b'"97975_20190725","103124"\n'
2 :  b'"85986_20161011","103114"\n'
3 :  b'"95754_20190327","103124"\n'
4 :  b'"77005_20170429","103120"\n'

data/ins.csv.bz2
0 :  b'"iid","date","score","type"\n'
1 :  b'"100010_20190329","03/29/2019 12:00:00 AM","-1","New Construction"\n'
2 :  b'"100010_20190403","04/03/2019 12:00:00 AM","100","Routine - Unscheduled"\n'
3 :  b'"100017_20190417","04/17/2019 12:00:00 AM","-1","New Ownership"\n'
4 :  b'"100017_20190816","08/16/2019 12:00:00 AM","91","Routine - Unscheduled"\n'

data/bus.csv.bz2
0 :  b'"business id column","name","address","city","state","postal_code","latitude","longitude","phone_number"\n'
1 :  b'"1000","HEUNG YUEN RESTAURANT","3279 22nd St","San Francisco","CA","94110","37.755282","-122.420493","-9999"\n'
2 :  b'"100010","ILLY CAFFE SF_PIER 39","PIER 39  K-106-B","San Francisco","CA","94133","-9999","-9999","+14154827284"\n'
3 :  b'"100017","AMICI\'S EAST COAST PIZZERIA","475 06th St","San F

<br/><br/><br/>

---

## Question 1c: Thinking about the files

Answer the following questions by filling in the correct boolean values in the following variables:

1. The bus.csv.bz2 file appears to be tab delimited.
1. The values all appear to be quoted.


<!--
BEGIN QUESTION
name: q1c
points: 1
-->

In [151]:
# True or False: The bus.csv.bz2 file appears to be tab delimited.
q1c_1 = False

# True or False: The values all appear to be quoted.
q1c_2 = True

In [152]:
ok.grade("q1c");

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Running tests

---------------------------------------------------------------------
Test summary
    Passed: 2
    Failed: 0
[ooooooooook] 100.0% passed



<br/><br/><br/>

---

## Question 1d: Reading in the Files

Based on the above information, let's attempt to load `bus.csv`, `ins2vio.csv`, `vio.csv` and `ins.csv` into Pandas dataframes with the following names: `bus`, `ins2vio`, `vio`, `ins` respectively.


<!--
BEGIN QUESTION
name: q1d
points: 1
-->

In [153]:
# path to directory containing data
bus = pd.read_csv("data/bus.csv.bz2")
ins = pd.read_csv("data/ins.csv.bz2")
ins2vio = pd.read_csv("data/ins2vio.csv.bz2")
vio = pd.read_csv("data/vio.csv.bz2")

In [154]:
ok.grade("q1d");

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Running tests

---------------------------------------------------------------------
Test summary
    Passed: 4
    Failed: 0
[ooooooooook] 100.0% passed



Now that you've read in the files, you can try some `pd.DataFrame` methods ([docs](https://pandas.pydata.org/pandas-docs/version/0.21/generated/pandas.DataFrame.html)).
You can use the `DataFrame.head` method to show the top few lines of the `bus`, `ins`, `ins2vio` and `vio` dataframes. 

In [155]:
bus.head()

business id column                         name                 address  \
0                1000        HEUNG YUEN RESTAURANT            3279 22nd St   
1              100010        ILLY CAFFE SF_PIER 39        PIER 39  K-106-B   
2              100017  AMICI'S EAST COAST PIZZERIA             475 06th St   
3              100026               LOCAL CATERING        1566 CARROLL AVE   
4              100030             OUI OUI! MACARON  2200 JERROLD AVE STE C   

            city state postal_code     latitude    longitude  phone_number  
0  San Francisco    CA       94110    37.755282  -122.420493         -9999  
1  San Francisco    CA       94133 -9999.000000 -9999.000000   14154827284  
2  San Francisco    CA       94103 -9999.000000 -9999.000000   14155279839  
3  San Francisco    CA       94124 -9999.000000 -9999.000000   14155860315  
4  San Francisco    CA       94124 -9999.000000 -9999.000000   14159702675

In [156]:
ins.head()

iid                    date  score                   type
0  100010_20190329  03/29/2019 12:00:00 AM     -1       New Construction
1  100010_20190403  04/03/2019 12:00:00 AM    100  Routine - Unscheduled
2  100017_20190417  04/17/2019 12:00:00 AM     -1          New Ownership
3  100017_20190816  08/16/2019 12:00:00 AM     91  Routine - Unscheduled
4  100017_20190826  08/26/2019 12:00:00 AM     -1  Reinspection/Followup

In [157]:
ins2vio.head()

iid     vid
0  97975_20190725  103124
1  85986_20161011  103114
2  95754_20190327  103124
3  77005_20170429  103120
4   4794_20181030  103138

In [158]:
vio.head()

description  risk_category     vid
0  Consumer advisory not provided for raw or unde...  Moderate Risk  103128
1                   Contaminated or adulterated food      High Risk  103108
2          Discharge from employee nose mouth or eye  Moderate Risk  103117
3                         Employee eating or smoking  Moderate Risk  103118
4                             Food in poor condition  Moderate Risk  103123

<br/><br/><br/>

---

## Question 1e: Identifying Issues with the Data

Use the `head` command on your four files again. This time, describe at least one potential problem with the data you see. Consider issues with missing values and bad data.

Please write your answer in the markdown cell below. You may create new cells below your answer to run code, but **please never add cells between a question cell and the answer cell below it.**

<!--
BEGIN QUESTION
name: q1e
manual: True
points: 1
-->
<!-- EXPORT TO PDF -->

*The problems come from these files:*
1. ins.csv.bz2

*There exists score column that has -1 as the value. The minimum value for a score is 0 (really bad score), not -1, up to 100 (best score). The -1 value can affect the average score for a certain location if we want to take the average score.*

2. bus.csv.bz2 

*There exists a negative phone number (-9999) which is not possible since there is no negative phone number. Then, the negative -9999.00 for the latitude and longitude are not possible for 2 different postal codes to have the same latitude and longitude.*

In [159]:
# Show work here.
ins.head()

iid                    date  score                   type
0  100010_20190329  03/29/2019 12:00:00 AM     -1       New Construction
1  100010_20190403  04/03/2019 12:00:00 AM    100  Routine - Unscheduled
2  100017_20190417  04/17/2019 12:00:00 AM     -1          New Ownership
3  100017_20190816  08/16/2019 12:00:00 AM     91  Routine - Unscheduled
4  100017_20190826  08/26/2019 12:00:00 AM     -1  Reinspection/Followup

In [160]:
bus.head()

business id column                         name                 address  \
0                1000        HEUNG YUEN RESTAURANT            3279 22nd St   
1              100010        ILLY CAFFE SF_PIER 39        PIER 39  K-106-B   
2              100017  AMICI'S EAST COAST PIZZERIA             475 06th St   
3              100026               LOCAL CATERING        1566 CARROLL AVE   
4              100030             OUI OUI! MACARON  2200 JERROLD AVE STE C   

            city state postal_code     latitude    longitude  phone_number  
0  San Francisco    CA       94110    37.755282  -122.420493         -9999  
1  San Francisco    CA       94133 -9999.000000 -9999.000000   14154827284  
2  San Francisco    CA       94103 -9999.000000 -9999.000000   14155279839  
3  San Francisco    CA       94124 -9999.000000 -9999.000000   14155860315  
4  San Francisco    CA       94124 -9999.000000 -9999.000000   14159702675

<br/><br/><br/>

---

<br/><br/><br/>

# 2: Examining the Business Data File

From its name alone, we expect the `bus.csv` file to contain information about the restaurants. Let's investigate the granularity of this dataset.

In [161]:
bus.head()

business id column                         name                 address  \
0                1000        HEUNG YUEN RESTAURANT            3279 22nd St   
1              100010        ILLY CAFFE SF_PIER 39        PIER 39  K-106-B   
2              100017  AMICI'S EAST COAST PIZZERIA             475 06th St   
3              100026               LOCAL CATERING        1566 CARROLL AVE   
4              100030             OUI OUI! MACARON  2200 JERROLD AVE STE C   

            city state postal_code     latitude    longitude  phone_number  
0  San Francisco    CA       94110    37.755282  -122.420493         -9999  
1  San Francisco    CA       94133 -9999.000000 -9999.000000   14154827284  
2  San Francisco    CA       94103 -9999.000000 -9999.000000   14155279839  
3  San Francisco    CA       94124 -9999.000000 -9999.000000   14155860315  
4  San Francisco    CA       94124 -9999.000000 -9999.000000   14159702675

<br/><br/><br/>

---

## Question 2a

The `bus` dataframe contains a column called `business id column` which probably corresponds to a unique business id.  However, let's first rename that column to `bid`.  Modify the `bus` dataframe by renaming that column to `bid`.

**Note**: In practice we might want to do this renaming when the table is loaded but for grading purposes we will do it here.

<!--
BEGIN QUESTION
name: q2a
points: 1
-->

In [162]:
bus = bus.rename({"business id column":"bid"}, axis = 'columns')
bus.head()

bid                         name                 address           city  \
0    1000        HEUNG YUEN RESTAURANT            3279 22nd St  San Francisco   
1  100010        ILLY CAFFE SF_PIER 39        PIER 39  K-106-B  San Francisco   
2  100017  AMICI'S EAST COAST PIZZERIA             475 06th St  San Francisco   
3  100026               LOCAL CATERING        1566 CARROLL AVE  San Francisco   
4  100030             OUI OUI! MACARON  2200 JERROLD AVE STE C  San Francisco   

  state postal_code     latitude    longitude  phone_number  
0    CA       94110    37.755282  -122.420493         -9999  
1    CA       94133 -9999.000000 -9999.000000   14154827284  
2    CA       94103 -9999.000000 -9999.000000   14155279839  
3    CA       94124 -9999.000000 -9999.000000   14155860315  
4    CA       94124 -9999.000000 -9999.000000   14159702675

In [163]:
ok.grade("q2a");

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Running tests

---------------------------------------------------------------------
Test summary
    Passed: 1
    Failed: 0
[ooooooooook] 100.0% passed



<br/><br/><br/>

---

## Question 2b

Examining the entries in `bus`, is the `bid` unique for each record (i.e. each row of data)? Your code should compute the answer, i.e. don't just hard code `True` or `False`.

Hint: use `value_counts()` or `unique()` to determine if the `bid` series has any duplicates.

<!--
BEGIN QUESTION
name: q2b
points: 1
-->

In [164]:
is_bid_unique = all(bus['bid'].value_counts() == 1)
is_bid_unique

True

In [165]:
ok.grade("q2b");

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Running tests

---------------------------------------------------------------------
Test summary
    Passed: 1
    Failed: 0
[ooooooooook] 100.0% passed



<br/><br/><br/>

---


## Question 2c

In the two cells below create two **series** 

1. where the index is the `name` of the business and the value is the number of records with that `name`
2. where the index is the `address` of the business and the value is the number of records with that `address`

Order both series in descending order by count. You may need to use `groupby()`, `size()`, `sort_values()`, or `value_counts()`. 

**Step 1**

<!--
BEGIN QUESTION
name: q2ci
points: 1
-->

In [166]:
name_counts = bus.groupby('name').size().sort_values(ascending = False)
#name_counts = bus['name'].value_counts().sort_values(ascending = False)
name_counts.head(20)

name
Peet's Coffee & Tea                                       20
Starbucks Coffee                                          13
Jamba Juice                                               10
McDonald's                                                10
STARBUCKS                                                  9
Proper Food                                                9
Mixt Greens/Mixt                                           8
Specialty's Cafe & Bakery                                  8
Philz Coffee                                               7
Starbucks                                                  7
Blue Bottle Coffee                                         7
The Organic Coup                                           7
Whole Foods Market                                         7
BlueStar Refreshment Services @ Uber Technologies, Inc     6
Lee's Deli                                                 6
Bon Appetit @ Twitter                                      6
JW Marriott SF Unio

In [167]:
ok.grade("q2ci");

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Running tests

---------------------------------------------------------------------
Test summary
    Passed: 2
    Failed: 0
[ooooooooook] 100.0% passed



**Step 2**

<!--
BEGIN QUESTION
name: q2cii
points: 1
-->

In [168]:
address_counts = bus.groupby('address').size().sort_values(ascending = False)
address_counts.head(10)

address
Off The Grid              39
428 11th St               34
3251 20th Ave             17
2948 Folsom St            17
Pier 41                   16
103 Horne Ave             14
24 Willie Mays Plaza      13
Off the Grid              11
1 United Nations Plaza    10
2948 Folsom St.           10
dtype: int64

In [169]:
ok.grade("q2cii");

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Running tests

---------------------------------------------------------------------
Test summary
    Passed: 2
    Failed: 0
[ooooooooook] 100.0% passed



<br/><br/><br/>

---


## Question 2d

Based on the above calculations answer each of the following questions by filling the value in the variable.

1. What does each record represent?  
1. What is the minimal primary key?

<!--
BEGIN QUESTION
name: q2d
points: 2
-->

In [170]:
# What does each record represent?  Valid answers are:
#    "One location of a restaurant."
#    "A chain of restaurants."
#    "A city block."
q2d_part1 = "One location of a restaurant."

# What is the minimal primary key? Valid answers are:
#    "bid"
#    "bid, name"
#    "bid, name, address"
q2d_part2 = "bid"

In [171]:
ok.grade("q2d");

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Running tests

---------------------------------------------------------------------
Test summary
    Passed: 2
    Failed: 0
[ooooooooook] 100.0% passed



<br/><br/><br/>

---

<br/><br/><br/>

# 3: Cleaning the Business Data Postal Codes

The business data contains postal code information that we can use to aggregate the ratings over regions of the city.  Let's examine and clean the postal code field.  The postal code (sometimes also called a ZIP code) partitions the city into regions:

<img src="https://www.usmapguide.com/wp-content/uploads/2019/03/printable-san-francisco-zip-code-map.jpg" alt="ZIP Code Map" style="width: 600px">

<br/><br/><br/>

---


## Question 3a

How many restaurants are in each ZIP code? 

In the cell below, create a **series** where the index is the postal code and the value is the number of records with that postal code in descending order of count. You may need to use `groupby()`, `size()`, or `value_counts()`. Do you notice any odd/invalid zip codes?

<!--
BEGIN QUESTION
name: q3a
points: 1
-->

In [172]:
zip_counts = bus.groupby('postal_code').size().sort_values(ascending = False)
print(zip_counts.to_string())

postal_code
94103         562
94110         555
94102         456
94107         408
94133         398
94109         382
94111         259
94122         255
94105         249
94118         231
94115         230
94108         229
94124         218
94114         200
-9999         194
94112         192
94117         189
94123         177
94121         157
94104         142
94132         132
94116          97
94158          90
94134          82
94127          67
94131          49
94130           8
94143           5
94013           2
94188           2
CA              2
94301           2
94101           2
95122           1
941033148       1
95133           1
95132           1
94102-5917      1
94014           1
941             1
94080           1
94105-2907      1
92672           1
64110           1
00000           1
94105-1420      1
941102019       1
95117           1
95112           1
95109           1
95105           1
94901           1
94621           1
94602           1
94544           

In [173]:
ok.grade("q3a");

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Running tests

---------------------------------------------------------------------
Test summary
    Passed: 3
    Failed: 0
[ooooooooook] 100.0% passed



<br/><br/><br/>

--- 

## Question 3b

Answer the following questions about the `postal_code` column in the `bus` dataframe.

1. The ZIP code column is which of the following type of data:
    1. Quantitative Continuous
    1. Quantitative Discrete
    1. Qualitative Ordinal
    1. Qualitative Nominal    
1. What Python data type is used to represent a ZIP code?

*Note*: ZIP codes and postal codes are the same thing.

Please write your answers in the variables below:

<!--
BEGIN QUESTION
name: q3b
points: 2
-->

In [174]:
# The ZIP code column is which of the following type of data:
#   "Quantitative Continuous" 
#   "Quantitative Discrete"
#   "Qualitative Ordinal"
#   "Qualitative Nominal"
q3b_part1 = "Qualitative Nominal"

# What Python data type is used to represent a ZIP code? 
#    "str"
#    "int"
#    "bool"
#    "float"
q3b_part2 = "str"

In [175]:
ok.grade("q3b");

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Running tests

---------------------------------------------------------------------
Test summary
    Passed: 2
    Failed: 0
[ooooooooook] 100.0% passed



<br/><br/><br/>

--- 

## Question 3c

In question 3a we noticed a large number of potentially invalid ZIP codes (e.g., "CA").  These are likely due to data entry errors.  To get a better understanding of the potential errors in the zip codes we will:

1. Import a list of valid San Francisco ZIP codes by using `pd.read_json` to load the file `data/sf_zipcodes.json` and extract a **series** of type `str` containing the valid ZIP codes.  *Hint: set `dtype` when invoking `read_json`.*
1. Construct a DataFrame containing only the businesses which DO NOT have valid ZIP codes.  You will probably want to use the `Series.isin` function. 


**Step 1**

<!--
BEGIN QUESTION
name: q3ci
points: 1
-->

In [176]:
valid_zips = pd.read_json("data/sf_zipcodes.json", dtype = str)['zip_codes']
valid_zips.head()

0    94102
1    94103
2    94104
3    94105
4    94107
Name: zip_codes, dtype: object

In [177]:
ok.grade("q3ci");

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Running tests

---------------------------------------------------------------------
Test summary
    Passed: 2
    Failed: 0
[ooooooooook] 100.0% passed



**Step 2**

<!--
BEGIN QUESTION
name: q3cii
points: 1
-->

In [178]:
invalid_zip_bus = bus[~bus['postal_code'].isin(valid_zips)]
invalid_zip_bus.head(20)

bid                               name  \
22   100126          Lamas Peruvian Food Truck   
68   100417                   COMPASS ONE, LLC   
96   100660                          TEAPENTER   
109  100781                  LE CAFE DU SOLEIL   
144  101084                     Deli North 200   
156  101129                    Vendor Room 200   
177  101192                       Cochinita #2   
276  102014       DROPBOX (Section 3, Floor 7)   
295  102245         Vessell CA Operations (#4)   
298   10227                   The Napper Tandy   
320   10372  BERNAL HEIGHTS NEIGBORHOOD CENTER   
321   10373                    El Tonayense #1   
322   10376                Good Frikin Chicken   
324   10406              Sunset Youth Services   
357   11416                   El Beach Burrito   
381   12199                      El Gallo Giro   
384   12344         The Village Market & Pizza   
406   13062              Everett Middle School   
434   13753                             Taboun   
548   17423                  Project Open Hand   

                                 address           city state postal_code  \
22                      Private Location  San Francisco    CA       -9999   
68                       1 MARKET ST. FL  San Francisco    CA  94105-1420   
96                        1518 IRVING ST  San Francisco    CA  94122-1909   
109                      200 FILLMORE ST  San Francisco    CA  94117-3504   
144  1 Warriors Way Level 300 North East  San Francisco    CA       94518   
156  1 Warriors Way Level 300 South West  San Francisco    CA       -9999   
177             2 Marina Blvd Fort Mason  San Francisco    CA       -9999   
276                        1800 Owens St  San Francisco    CA       -9999   
295                      2351 Mission St  San Francisco    CA       -9999   
298                         3200 24th St  San Francisco    CA       -9999   
320                     515 CORTLAND AVE  San Francisco    CA       -9999   
321                     1717 Harrison St  San Francisco    CA       -9999   
322                           10 29th St  San Francisco    CA       -9999   
324                        3918 Judah St  San Francisco    CA       -9999   
357                        3914 Judah St  San Francisco    CA       -9999   
381                         3055 23rd St  San Francisco    CA       -9999   
384                        750 Font Blvd  San Francisco    CA       -9999   
406                        450 Church St  San Francisco    CA       -9999   
434                    203 Parnassus Ave  San Francisco    CA       -9999   
548                       100 Diamond St  San Francisco    CA       -9999   

        latitude    longitude  phone_number  
22  -9999.000000 -9999.000000         -9999  
68  -9999.000000 -9999.000000   14154324000  
96  -9999.000000 -9999.000000   14155868318  
109 -9999.000000 -9999.000000   14155614215  
144 -9999.000000 -9999.000000         -9999  
156 -9999.000000 -9999.000000         -9999  
177 -9999.000000 -9999.000000   14150429222  
276 -9999.000000 -9999.000000         -9999  
295 -9999.000000 -9999.000000         -9999  
298    37.752581  -122.416482         -9999  
320    37.739110  -122.416404   14155202142  
321    37.769426  -122.413446   14155556127  
322    37.744369  -122.420967         -9999  
324    37.760560  -122.504027         -9999  
357    37.760851  -122.503998         -9999  
381    37.754218  -122.413285   14155553048  
384    37.723462  -122.483012   14155374525  
406    37.763794  -122.428617         -9999  
434    37.764574  -122.452950         -9999  
548    37.760689  -122.437252         -9999

In [179]:
ok.grade("q3cii");

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Running tests

---------------------------------------------------------------------
Test summary
    Passed: 2
    Failed: 0
[ooooooooook] 100.0% passed



<br/><br/><br/>

--- 

## Question 3d

In the previous question, many of the businesses had a common invalid postal code that was likely used to code a MISSING postal code.  Do they all share a potentially "interesting address"?

In the following cell, construct a **series** that counts the number of businesses at each `address` that have this single likely MISSING postal code value.  Order the series in descending order by count. 

After examining the output.  Answer the following question by filling in the appropriate variable. If we were to drop businesses with MISSING postal code values would a particular class of business be affected?  If you are unsure try to search the web for the most common addresses.


<!--
BEGIN QUESTION
name: q3d
points: 3
-->

In [180]:
zip_9999 = invalid_zip_bus[invalid_zip_bus['postal_code'] == '-9999']
missing_zip_address_count = zip_9999.groupby('address').size().sort_values(ascending = False)
missing_zip_address_count.head()

address
Off The Grid                  39
Off the Grid                  10
OTG                            4
Approved Locations             3
Approved Private Locations     3
dtype: int64

In [181]:
ok.grade("q3d");

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Running tests

---------------------------------------------------------------------
Test summary
    Passed: 3
    Failed: 0
[ooooooooook] 100.0% passed



<br/><br/><br/>

--- 

## Question 3e

**True or False**:  *If we were to drop businesses with MISSING postal code values a particular class of business will be affected.*

<!--
BEGIN QUESTION
name: q3e
points: 2
-->

In [182]:
# True or False: 
#  If we were to drop businesses with MISSING postal code values 
#   a particular class of business be affected.
q3d_true_or_false = True

In [183]:
ok.grade("q3e");

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Running tests

---------------------------------------------------------------------
Test summary
    Passed: 1
    Failed: 0
[ooooooooook] 100.0% passed



<br/><br/><br/>

--- 

## Question 3f

Examine the `invalid_zip_bus` dataframe we computed above and look at the businesses that DO NOT have the special MISSING ZIP code value.  Some of the invalid postal codes are just the full 9 digit code rather than the first 5 digits.  Create a new column named `postal5` in the original `bus` dataframe which contains only the first 5 digits of the `postal_code` column.   Finally, for any of the `postal5` ZIP code entries that were not a valid San Fransisco ZIP Code (according to `valid_zips`) set the entry to `None`.  


<!--
BEGIN QUESTION
name: q3f
points: 2
-->

In [184]:
bus['postal5'] = None
first5_postal = bus['postal_code'].str[:5]
bus.loc[bus[first5_postal.isin(valid_zips)].index, 'postal5'] = first5_postal

# Checking the corrected postal5 column
bus.loc[invalid_zip_bus.index, ['bid', 'name', 'postal_code', 'postal5']]

bid                            name postal_code postal5
22    100126       Lamas Peruvian Food Truck       -9999    None
68    100417                COMPASS ONE, LLC  94105-1420   94105
96    100660                       TEAPENTER  94122-1909   94122
109   100781               LE CAFE DU SOLEIL  94117-3504   94117
144   101084                  Deli North 200       94518    None
...      ...                             ...         ...     ...
6173   99369                     HOTEL BIRON  94102-5917   94102
6174   99376  Mashallah Halal Food truck Ind       -9999    None
6199   99536               FAITH SANDWICH #2  94105-2907   94105
6204   99681                         Twister       95112    None
6241   99819                  CHESTNUT DINER  94123-3106   94123

[230 rows x 4 columns]

In [185]:
ok.grade("q3f");

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Running tests

---------------------------------------------------------------------
Test summary
    Passed: 4
    Failed: 0
[ooooooooook] 100.0% passed



<br/><br/><br/>

---

<br/><br/><br/>

# 4: Investigate the Inspection Data

Let's now turn to the inspection DataFrame. Earlier, we found that `ins` has 4 columns named 
`iid`, `score`, `date` and `type`.  In this section, we determine the granularity of `ins` and investigate the kinds of information provided for the inspections. 

Let's start by looking again at the first 5 rows of `ins` to see what we're working with.

In [186]:
ins.head(5)

iid                    date  score                   type
0  100010_20190329  03/29/2019 12:00:00 AM     -1       New Construction
1  100010_20190403  04/03/2019 12:00:00 AM    100  Routine - Unscheduled
2  100017_20190417  04/17/2019 12:00:00 AM     -1          New Ownership
3  100017_20190816  08/16/2019 12:00:00 AM     91  Routine - Unscheduled
4  100017_20190826  08/26/2019 12:00:00 AM     -1  Reinspection/Followup

<br/><br/><br/>

---

## Question 4a

The column `iid` probably corresponds to an inspection id.  Is it a primary key?  Write an expression (line of code) that evaluates to 'True' or 'False' based on whether all the values are unique.


<!--
BEGIN QUESTION
name: q4a
points: 1
-->

In [187]:
is_ins_iid_a_primary_key = all(ins['iid'].value_counts() == 1)
is_ins_iid_a_primary_key

True

In [188]:
ok.grade("q4a");

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Running tests

---------------------------------------------------------------------
Test summary
    Passed: 1
    Failed: 0
[ooooooooook] 100.0% passed



<br/><br/><br/>

---

## Question 4b

The column `iid` appears to be the composition of two numbers and the first number looks like a business id.  

**Part 1.**: Create a new column called `bid` in the `ins` dataframe containing just the business id.  You will want to use `ins['iid'].str` operations to do this.  Also be sure to convert the type of this column to `int`

**Part 2.**: Then compute how many values in this new column are also valid business ids (appear in the `bus['bid']` column). This is verifying a foreign key relationship. Consider using the `pd.Series.isin` function.

**Part 3.**: Answer True or False, `ins['bid']` is a foreign key reference to `bus['bid']`.


**No python `for` loops or list comprehensions required!**

**Part 1**

<!--
BEGIN QUESTION
name: q4bi
points: 1
-->

In [189]:
ins['bid'] = ins['iid'].str.split('_').str[0].astype(int)
ins.head()

iid                    date  score                   type  \
0  100010_20190329  03/29/2019 12:00:00 AM     -1       New Construction   
1  100010_20190403  04/03/2019 12:00:00 AM    100  Routine - Unscheduled   
2  100017_20190417  04/17/2019 12:00:00 AM     -1          New Ownership   
3  100017_20190816  08/16/2019 12:00:00 AM     91  Routine - Unscheduled   
4  100017_20190826  08/26/2019 12:00:00 AM     -1  Reinspection/Followup   

      bid  
0  100010  
1  100010  
2  100017  
3  100017  
4  100017

In [190]:
ok.grade("q4bi");

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Running tests

---------------------------------------------------------------------
Test summary
    Passed: 3
    Failed: 0
[ooooooooook] 100.0% passed



**Part 2**

<!--
BEGIN QUESTION
name: q4bii
points: 1
-->

In [191]:
invalid_bid_count = len(ins[~ins['bid'].isin(bus['bid'])])
invalid_bid_count

0

In [192]:
ok.grade("q4bii");

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Running tests

---------------------------------------------------------------------
Test summary
    Passed: 1
    Failed: 0
[ooooooooook] 100.0% passed



**Part 3**

<!--
BEGIN QUESTION
name: q4biii
points: 1
-->

In [193]:
# True or False: The column ins['bid'] is a foreign key 
#   referencing the bus['bid'] primary key.

q4b_is_foreign_key = True

In [194]:
ok.grade("q4biii");

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Running tests

---------------------------------------------------------------------
Test summary
    Passed: 1
    Failed: 0
[ooooooooook] 100.0% passed



<br/><br/><br/>

---

## Question 4c

What if we are interested in a time component of the inspection data?  We need to examine the date column of each inspection. 

**Part 1:** What is the type of the individual `ins['date']` entries. You may want to grab the very first entry and use the `type` function in python. 

**Part 2:** Use `pd.to_datetime` to create a new `ins['timestamp']` column containing of `pd.Timestamp` objects.  These will allow us to do more date manipulation.

**Part 3:** What are the earliest and latest dates in our inspection data?  *Hint: you can use `min` and `max` on dates of the correct type.*

**Part 4:** We probably want to examine the inspections by year. Create an additional `ins['year']` column containing just the year of the inspection.  Consider using `pd.Series.dt.year` to do this.

**No python `for` loops or list comprehensions required!**

**Part 1**

<!--
BEGIN QUESTION
name: q4ci
points: 1
-->

In [195]:
ins_date_type = type(ins['date'])
ins_date_type

pandas.core.series.Series

In [196]:
ok.grade("q4ci");

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Running tests

---------------------------------------------------------------------
Test summary
    Passed: 1
    Failed: 0
[ooooooooook] 100.0% passed



**Part 2**

<!--
BEGIN QUESTION
name: q4cii
points: 1
-->

In [197]:
ins['timestamp'] = pd.to_datetime(ins['date'])
ins.head()

iid                    date  score                   type  \
0  100010_20190329  03/29/2019 12:00:00 AM     -1       New Construction   
1  100010_20190403  04/03/2019 12:00:00 AM    100  Routine - Unscheduled   
2  100017_20190417  04/17/2019 12:00:00 AM     -1          New Ownership   
3  100017_20190816  08/16/2019 12:00:00 AM     91  Routine - Unscheduled   
4  100017_20190826  08/26/2019 12:00:00 AM     -1  Reinspection/Followup   

      bid  timestamp  
0  100010 2019-03-29  
1  100010 2019-04-03  
2  100017 2019-04-17  
3  100017 2019-08-16  
4  100017 2019-08-26

In [198]:
ok.grade("q4cii");

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Running tests

---------------------------------------------------------------------
Test summary
    Passed: 1
    Failed: 0
[ooooooooook] 100.0% passed



**Part 3**

<!--
BEGIN QUESTION
name: q4ciii
points: 1
-->

In [199]:
earliest_date = min(ins['timestamp'])
latest_date = max(ins['timestamp'])

print("Earliest Date:", earliest_date)
print("Latest Date:", latest_date)

Earliest Date: 2016-10-04 00:00:00
Latest Date: 2019-11-28 00:00:00


In [200]:
ok.grade("q4ciii");

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Running tests

---------------------------------------------------------------------
Test summary
    Passed: 2
    Failed: 0
[ooooooooook] 100.0% passed



**Part 4**

<!--
BEGIN QUESTION
name: q4civ
points: 1
-->

In [201]:
ins['year'] = ins['timestamp'].dt.year

In [202]:
ok.grade("q4civ");

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Running tests

---------------------------------------------------------------------
Test summary
    Passed: 1
    Failed: 0
[ooooooooook] 100.0% passed



In [203]:
ins.head()

iid                    date  score                   type  \
0  100010_20190329  03/29/2019 12:00:00 AM     -1       New Construction   
1  100010_20190403  04/03/2019 12:00:00 AM    100  Routine - Unscheduled   
2  100017_20190417  04/17/2019 12:00:00 AM     -1          New Ownership   
3  100017_20190816  08/16/2019 12:00:00 AM     91  Routine - Unscheduled   
4  100017_20190826  08/26/2019 12:00:00 AM     -1  Reinspection/Followup   

      bid  timestamp  year  
0  100010 2019-03-29  2019  
1  100010 2019-04-03  2019  
2  100017 2019-04-17  2019  
3  100017 2019-08-16  2019  
4  100017 2019-08-26  2019

<br/><br/><br/>

---

## Question 4d

What is the relationship between the type of inspection over the 2016 to 2019 timeframe? 

**Part 1**

Construct the following table by
1. Using the `pivot_table` containing the number (`size`) of inspections for the given `type` and `year`.
1. Adding an extra `Total` column to the result using `sum`
1. Sort the results in descending order by the `Total`.

<table border="1" class="dataframe">  <thead>    <tr style="text-align: right;">      <th>year</th>      <th>2016</th>      <th>2017</th>      <th>2018</th>      <th>2019</th>      <th>Total</th>    </tr>    <tr>      <th>type</th>      <th></th>      <th></th>      <th></th>      <th></th>      <th></th>    </tr>  </thead>  <tbody>    <tr>      <th>Routine - Unscheduled</th>      <td>966</td>      <td>4057</td>      <td>4373</td>      <td>4681</td>      <td>14077</td>    </tr>    <tr>      <th>Reinspection/Followup</th>      <td>445</td>      <td>1767</td>      <td>1935</td>      <td>2292</td>      <td>6439</td>    </tr>    <tr>      <th>New Ownership</th>      <td>99</td>      <td>506</td>      <td>528</td>      <td>459</td>      <td>1592</td>    </tr>    <tr>      <th>Complaint</th>      <td>91</td>      <td>418</td>      <td>512</td>      <td>437</td>      <td>1458</td>    </tr>    <tr>      <th>New Construction</th>      <td>102</td>      <td>485</td>      <td>218</td>      <td>189</td>      <td>994</td>    </tr>    <tr>      <th>Non-inspection site visit</th>      <td>51</td>      <td>276</td>      <td>253</td>      <td>231</td>      <td>811</td>    </tr>    <tr>      <th>New Ownership - Followup</th>      <td>0</td>      <td>45</td>      <td>219</td>      <td>235</td>      <td>499</td>    </tr>    <tr>      <th>Structural Inspection</th>      <td>1</td>      <td>153</td>      <td>50</td>      <td>190</td>      <td>394</td>    </tr>    <tr>      <th>Complaint Reinspection/Followup</th>      <td>19</td>      <td>68</td>      <td>70</td>      <td>70</td>      <td>227</td>    </tr>    <tr>      <th>Foodborne Illness Investigation</th>      <td>1</td>      <td>29</td>      <td>50</td>      <td>35</td>      <td>115</td>    </tr>    <tr>      <th>Routine - Scheduled</th>      <td>0</td>      <td>9</td>      <td>8</td>      <td>29</td>      <td>46</td>    </tr>    <tr>      <th>Administrative or Document Review</th>      <td>2</td>      <td>1</td>      <td>1</td>      <td>0</td>      <td>4</td>    </tr>    <tr>      <th>Multi-agency Investigation</th>      <td>0</td>      <td>0</td>      <td>1</td>      <td>2</td>      <td>3</td>    </tr>    <tr>      <th>Special Event</th>      <td>0</td>      <td>3</td>      <td>0</td>      <td>0</td>      <td>3</td>    </tr>    <tr>      <th>Community Health Assessment</th>      <td>1</td>      <td>0</td>      <td>0</td>      <td>0</td>      <td>1</td>    </tr>  </tbody></table>


**Part 2**

Based on the above analysis, which year appears to have had a lot of businesses in new buildings?

**No python `for` loops or list comprehensions required!**

**Part 1**

<!--
BEGIN QUESTION
name: q4di
points: 2
-->

In [204]:
ins_pivot = pd.pivot_table(ins, values = 'bid', index = ['type'], 
                           columns = ['year'], aggfunc = np.size, fill_value = 0)
ins_pivot['Total'] = ins['type'].value_counts()

ins_pivot_sorted = ins_pivot.sort_values(by = ['Total'], ascending = False)
ins_pivot_sorted

year                               2016  2017  2018  2019  Total
type                                                            
Routine - Unscheduled               966  4057  4373  4681  14077
Reinspection/Followup               445  1767  1935  2292   6439
New Ownership                        99   506   528   459   1592
Complaint                            91   418   512   437   1458
New Construction                    102   485   218   189    994
Non-inspection site visit            51   276   253   231    811
New Ownership - Followup              0    45   219   235    499
Structural Inspection                 1   153    50   190    394
Complaint Reinspection/Followup      19    68    70    70    227
Foodborne Illness Investigation       1    29    50    35    115
Routine - Scheduled                   0     9     8    29     46
Administrative or Document Review     2     1     1     0      4
Multi-agency Investigation            0     0     1     2      3
Special Event                         0     3     0     0      3
Community Health Assessment           1     0     0     0      1

In [205]:
ok.grade("q4di");

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Running tests

---------------------------------------------------------------------
Test summary
    Passed: 2
    Failed: 0
[ooooooooook] 100.0% passed



**Part 2**

<!--
BEGIN QUESTION
name: q4dii
points: 2
-->

In [206]:
year_of_new_construction = 2017
year_of_new_construction

2017

In [207]:
ok.grade("q4dii");

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Running tests

---------------------------------------------------------------------
Test summary
    Passed: 1
    Failed: 0
[ooooooooook] 100.0% passed



<br/><br/><br/>

---

## Question 4e

Let's examine the inspection scores `ins['score']`



In [208]:
ins['score'].value_counts().head()

-1      12632
 100     1993
 96      1681
 92      1260
 94      1250
Name: score, dtype: int64

There are a large number of inspections with the `'score'` of `-1`.   These are probably missing values.  Let's see what type of inspections have scores and which do not.  Create the following dataframe using steps similar to the previous question.

You should observe that inspection scores appear only to be assigned to `Routine - Unscheduled` inspections.


<table border="1" class="dataframe">  <thead>    <tr style="text-align: right;">      <th>Missing Score</th>      <th>False</th>      <th>True</th>      <th>Total</th>    </tr>    <tr>      <th>type</th>      <th></th>      <th></th>      <th></th>    </tr>  </thead>  <tbody>    <tr>      <th>Routine - Unscheduled</th>      <td>14031</td>      <td>46</td>      <td>14077</td>    </tr>    <tr>      <th>Reinspection/Followup</th>      <td>0</td>      <td>6439</td>      <td>6439</td>    </tr>    <tr>      <th>New Ownership</th>      <td>0</td>      <td>1592</td>      <td>1592</td>    </tr>    <tr>      <th>Complaint</th>      <td>0</td>      <td>1458</td>      <td>1458</td>    </tr>    <tr>      <th>New Construction</th>      <td>0</td>      <td>994</td>      <td>994</td>    </tr>    <tr>      <th>Non-inspection site visit</th>      <td>0</td>      <td>811</td>      <td>811</td>    </tr>    <tr>      <th>New Ownership - Followup</th>      <td>0</td>      <td>499</td>      <td>499</td>    </tr>    <tr>      <th>Structural Inspection</th>      <td>0</td>      <td>394</td>      <td>394</td>    </tr>    <tr>      <th>Complaint Reinspection/Followup</th>      <td>0</td>      <td>227</td>      <td>227</td>    </tr>    <tr>      <th>Foodborne Illness Investigation</th>      <td>0</td>      <td>115</td>      <td>115</td>    </tr>    <tr>      <th>Routine - Scheduled</th>      <td>0</td>      <td>46</td>      <td>46</td>    </tr>    <tr>      <th>Administrative or Document Review</th>      <td>0</td>      <td>4</td>      <td>4</td>    </tr>    <tr>      <th>Multi-agency Investigation</th>      <td>0</td>      <td>3</td>      <td>3</td>    </tr>    <tr>      <th>Special Event</th>      <td>0</td>      <td>3</td>      <td>3</td>    </tr>    <tr>      <th>Community Health Assessment</th>      <td>0</td>      <td>1</td>      <td>1</td>    </tr>  </tbody></table>



<!--
BEGIN QUESTION
name: q4e
points: 2
-->

In [209]:
ins['Missing Score'] = (ins['score'] == -1)
ins_missing_score_pivot = pd.pivot_table(ins, values = 'score', index = ['type'], 
                                         columns = ['Missing Score'], 
                                         aggfunc = np.size, fill_value = 0)

ins_missing_score_pivot['Total'] = ins['type'].value_counts()
ins_missing_score_pivot = ins_missing_score_pivot.sort_values(by = ['Total'], ascending = False)
ins_missing_score_pivot

Missing Score                      False  True  Total
type                                                 
Routine - Unscheduled              14031    46  14077
Reinspection/Followup                  0  6439   6439
New Ownership                          0  1592   1592
Complaint                              0  1458   1458
New Construction                       0   994    994
Non-inspection site visit              0   811    811
New Ownership - Followup               0   499    499
Structural Inspection                  0   394    394
Complaint Reinspection/Followup        0   227    227
Foodborne Illness Investigation        0   115    115
Routine - Scheduled                    0    46     46
Administrative or Document Review      0     4      4
Multi-agency Investigation             0     3      3
Special Event                          0     3      3
Community Health Assessment            0     1      1

In [210]:
ok.grade("q4e");

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Running tests

---------------------------------------------------------------------
Test summary
    Passed: 1
    Failed: 0
[ooooooooook] 100.0% passed



Notice that inspection scores appear only to be assigned to `Routine - Unscheduled` inspections. It is reasonable that for inspection types such as `New Ownership` and `Complaint` to have no associated inspection scores, but we might be curious why there are no inspection scores for the `Reinspection/Followup` inspection type.

<br/><br/><br/>

---

<br/><br/><br/>

# 5: Joining Data Across Tables

In this question we will start to connect data across mulitple tables.  We will be using the `merge` function. 

<br/><br/><br/>

--- 

## Question 5a

Let's figure out which restaurants had the lowest scores. Let's start by creating a new dataframe called `ins_named`. It should be exactly the same as `ins`, except that it should have the name and address of every business, as determined by the `bus` dataframe. 

*Hint*: Use the merge method to join the `ins` dataframe with the appropriate portion of the `bus` dataframe. See the official [documentation](https://pandas.pydata.org/pandas-docs/stable/user_guide/merging.html) on how to use `merge`.

*Note*: For quick reference, a pandas 'left' join keeps the keys from the left frame, so if ins is the left frame, all the keys from ins are kept and if a set of these keys don't have matches in the other frame, the columns from the other frame for these "unmatched" key rows contains NaNs.

<!--
BEGIN QUESTION
name: q5a
points: 1
-->

In [211]:
ins_named = ins.merge(bus[['bid', 'name', 'address']], how = 'left', on = 'bid')
ins_named.head()

iid                    date  score                   type  \
0  100010_20190329  03/29/2019 12:00:00 AM     -1       New Construction   
1  100010_20190403  04/03/2019 12:00:00 AM    100  Routine - Unscheduled   
2  100017_20190417  04/17/2019 12:00:00 AM     -1          New Ownership   
3  100017_20190816  08/16/2019 12:00:00 AM     91  Routine - Unscheduled   
4  100017_20190826  08/26/2019 12:00:00 AM     -1  Reinspection/Followup   

      bid  timestamp  year  Missing Score                         name  \
0  100010 2019-03-29  2019           True        ILLY CAFFE SF_PIER 39   
1  100010 2019-04-03  2019          False        ILLY CAFFE SF_PIER 39   
2  100017 2019-04-17  2019           True  AMICI'S EAST COAST PIZZERIA   
3  100017 2019-08-16  2019          False  AMICI'S EAST COAST PIZZERIA   
4  100017 2019-08-26  2019           True  AMICI'S EAST COAST PIZZERIA   

            address  
0  PIER 39  K-106-B  
1  PIER 39  K-106-B  
2       475 06th St  
3       475 06th St  
4       475 06th St

In [212]:
ok.grade("q5a");

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Running tests

---------------------------------------------------------------------
Test summary
    Passed: 3
    Failed: 0
[ooooooooook] 100.0% passed



<br/><br/><br/>

--- 

## Question 5b

Let's look at the 20 businesses with the lowest **median** score.  Order your results by the median score followed by the business id to break ties. The resulting table should look like:


*Hint: You may find the `as_index` argument important*

<table border="1" class="dataframe">  <thead>    <tr style="text-align: right;">      <th></th>      <th>bid</th>      <th>name</th>      <th>median score</th>    </tr>  </thead>  <tbody>    <tr>      <th>3876</th>      <td>84590</td>      <td>Chaat Corner</td>      <td>54.0</td>    </tr>    <tr>      <th>4564</th>      <td>90622</td>      <td>Taqueria Lolita</td>      <td>57.0</td>    </tr>    <tr>      <th>4990</th>      <td>94351</td>      <td>VBowls LLC</td>      <td>58.0</td>    </tr>    <tr>      <th>2719</th>      <td>69282</td>      <td>New Jumbo Seafood Restaurant</td>      <td>60.5</td>    </tr>    <tr>      <th>222</th>      <td>1154</td>      <td>SUNFLOWER RESTAURANT</td>      <td>63.5</td>    </tr>    <tr>      <th>1991</th>      <td>39776</td>      <td>Duc Loi Supermarket</td>      <td>64.0</td>    </tr>    <tr>      <th>2734</th>      <td>69397</td>      <td>Minna SF Group LLC</td>      <td>64.0</td>    </tr>    <tr>      <th>3291</th>      <td>78328</td>      <td>Golden Wok</td>      <td>64.0</td>    </tr>    <tr>      <th>4870</th>      <td>93150</td>      <td>Chez Beesen</td>      <td>64.0</td>    </tr>    <tr>      <th>4911</th>      <td>93502</td>      <td>Smoky Man</td>      <td>64.0</td>    </tr>    <tr>      <th>5510</th>      <td>98995</td>      <td>Vallarta's Taco Bar</td>      <td>64.0</td>    </tr>    <tr>      <th>1457</th>      <td>10877</td>      <td>CHINA FIRST INC.</td>      <td>64.5</td>    </tr>    <tr>      <th>2890</th>      <td>71310</td>      <td>Golden King Vietnamese Restaurant</td>      <td>64.5</td>    </tr>    <tr>      <th>4352</th>      <td>89070</td>      <td>Lafayette Coffee Shop</td>      <td>64.5</td>    </tr>    <tr>      <th>505</th>      <td>2542</td>      <td>PETER D'S RESTAURANT</td>      <td>65.0</td>    </tr>    <tr>      <th>2874</th>      <td>71008</td>      <td>House of Pancakes</td>      <td>65.0</td>    </tr>    <tr>      <th>818</th>      <td>3862</td>      <td>IMPERIAL GARDEN SEAFOOD RESTAURANT</td>      <td>66.0</td>    </tr>    <tr>      <th>2141</th>      <td>61427</td>      <td>Nick's Foods</td>      <td>66.0</td>    </tr>    <tr>      <th>2954</th>      <td>72176</td>      <td>Wolfes Lunch</td>      <td>66.0</td>    </tr>    <tr>      <th>4367</th>      <td>89141</td>      <td>Cha Cha Cha on Mission</td>      <td>66.5</td>    </tr>  </tbody></table>


<!--
BEGIN QUESTION
name: q5b
points: 3
-->

In [213]:
no_missing_score = ins_named[ins_named['Missing Score'] == False].drop(columns = ['year', 'Missing Score'])
twenty_lowest_scoring = no_missing_score.groupby(['bid', 'name'],
                                                as_index = False).median().sort_values(['score', 'bid'])
twenty_lowest_scoring = twenty_lowest_scoring.rename({'score' : 'median score'}, axis = 'columns')[:20]
twenty_lowest_scoring


bid                                name  median score
3876  84590                        Chaat Corner          54.0
4564  90622                     Taqueria Lolita          57.0
4990  94351                          VBowls LLC          58.0
2719  69282        New Jumbo Seafood Restaurant          60.5
222    1154                SUNFLOWER RESTAURANT          63.5
1991  39776                 Duc Loi Supermarket          64.0
2734  69397                  Minna SF Group LLC          64.0
3291  78328                          Golden Wok          64.0
4870  93150                         Chez Beesen          64.0
4911  93502                           Smoky Man          64.0
5510  98995                 Vallarta's Taco Bar          64.0
1457  10877                    CHINA FIRST INC.          64.5
2890  71310   Golden King Vietnamese Restaurant          64.5
4352  89070               Lafayette Coffee Shop          64.5
505    2542                PETER D'S RESTAURANT          65.0
2874  71008                   House of Pancakes          65.0
818    3862  IMPERIAL GARDEN SEAFOOD RESTAURANT          66.0
2141  61427                        Nick's Foods          66.0
2954  72176                        Wolfes Lunch          66.0
4367  89141              Cha Cha Cha on Mission          66.5

In [214]:
ok.grade("q5b");

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Running tests

---------------------------------------------------------------------
Test summary
    Passed: 2
    Failed: 0
[ooooooooook] 100.0% passed



<br/><br/><br/>

--- 

## Question 5c

Let's now examine the descriptions of violations for inspections with `score > 0` and `score < 65`.  Construct a **Series** indexed by the `description` of the violation from the `vio` table with the value being the number of times that violation occured for inspections with the above score range.  Sort the results in descending order of the count.

The first few entries should look like:

```
Unclean or unsanitary food contact surfaces                                  43
High risk food holding temperature                                           42
Unclean or degraded floors walls or ceilings                                 40
Unapproved or unmaintained equipment or utensils                             39
```
You will need to use `merge` twice.

<!--
BEGIN QUESTION
name: q5c
points: 2
-->

In [215]:
merged_vio = vio.merge(ins2vio, how = 'left', 
                       on = 'vid').merge(ins[['iid', 'score']], how = 'left', on = 'iid')
merged_vio = merged_vio[merged_vio['score'] > 0]
merged_vio = merged_vio[merged_vio['score'] < 65]
merged_vio

low_score_violations = merged_vio.groupby('description').size().sort_values(ascending = False)
#merged_vio['description'].value_counts().sort_values(ascending = False)
low_score_violations.head(20)

description
Unclean or unsanitary food contact surfaces                                  43
High risk food holding temperature                                           42
Unclean or degraded floors walls or ceilings                                 40
Unapproved or unmaintained equipment or utensils                             39
Foods not protected from contamination                                       37
High risk vermin infestation                                                 37
Inadequate food safety knowledge or lack of certified food safety manager    35
Inadequate and inaccessible handwashing facilities                           35
Improper thawing methods                                                     30
Unclean hands or improper use of gloves                                      27
Improper cooling methods                                                     25
Unclean nonfood contact surfaces                                             21
Inadequately cleaned or sani

In [216]:
ok.grade("q5c");

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Running tests

---------------------------------------------------------------------
Test summary
    Passed: 3
    Failed: 0
[ooooooooook] 100.0% passed



<br/><br/><br/>

---

<br/><br/><br/>

# 6: Compute Something Interesting

Play with the data and try to compute something interesting about the data. Please try to use at least one of groupby, pivot, or merge (or all of the above).  

Please show your work in the cell below and describe in words what you found in the same cell. This question will be graded leniently but good solutions may be used to create future homework problems.

**Please have both your code and your explanation in the same one cell below. Any work in any other cell will not be graded.**


<!--
BEGIN QUESTION
name: q6
points: 3
manual: True
-->

<!-- EXPORT TO PDF -->

In [217]:
#YOUR CODE HERE
new_bus = bus
new_bus['Has PhoneNum'] = (new_bus['phone_number'] != -9999)

has_phone_pivot_table = pd.pivot_table(new_bus, values = 'bid', index = ['postal5'], 
                                       columns = ['Has PhoneNum'], aggfunc = np.size, 
                                       fill_value = 0, margins = True, margins_name = 'Total')
has_phone_pivot_table = has_phone_pivot_table.sort_values(by = ['Total'], ascending = False)
has_phone_pivot_table

# YOUR EXPLANATION HERE (in a comment)
# Here, I make a copy of bus dataframe called new_bus to 
# prevent changing the bus dataframe data. I make a new 
# column called 'Has PhoneNum' which have a value either
# the restaurant has a phone number (True or False).

# Then, I make a table with 'postal5' as the index and 
# 'Has PhoneNum' as the columns to know how many restaurants
# based on the postal code have phone numbers.

# Based on the table below, in San Francisco area, only 1989
# restaurants have a phone number, and the other 4043 do not.
# Then, based on the data, we can know that the resturants
# that do not provide phone number have an invalid phone
# number (-9999) based on our sorting in new_data.



Has PhoneNum  False  True  Total
postal5                         
Total          4043  1989   6032
94103           348   215    563
94110           287   269    556
94102           357   100    457
94107           329    79    408
94133           344    54    398
94109           291    91    382
94111           225    34    259
94122           136   120    256
94105           204    47    251
94118           147    84    231
94115           149    81    230
94108           179    50    229
94124           139    80    219
94114            85   115    200
94112           108    84    192
94117           103    87    190
94123           115    63    178
94121            91    66    157
94104           122    20    142
94132            42    90    132
94116            58    39     97
94158            76    14     90
94134            57    25     82
94127            22    45     67
94131            16    33     49
94130             7     1      8
94143             3     2      5
94188             1     1      2
94129             1     0      1
94120             1     0      1

In [218]:
#THIS CELL AND ANY CELLS ADDED BELOW WILL NOT BE GRADED

## Congratulations! You have finished Part 1 of Project 1! 

In our analysis of the business data, we found that there are some errors with the ZIP codes. As a result, we made the records with ZIP codes outside of San Francisco have a `postal5` value of `None` and shortened 9-digit zip codes to 5-digit ones. In practice, we could take the time to look up the restaurant address online and fix some of the zip code issues.

In our analysis of the inspection data, we investigated the relationship between the year and type of inspection, and we figured out that only `Routine - Unscheduled` inspections have inspection scores.
Finally, we joined the business and inspection data to identify restaurants with the worst ratings and the lowest median scores. 

# Submit
Make sure you have run all cells in your notebook in order before running the cell below, so that all images/graphs appear in the output.
**Please save before submitting!**

<!-- EXPECT 2 EXPORTED QUESTIONS -->

In [ ]:
# Save your notebook first, then run this cell to submit.
import jassign.to_pdf
jassign.to_pdf.generate_pdf('proj1a.ipynb', 'proj1a.pdf')
ok.submit()

Generating PDF...
Saved proj1a.pdf
